File: windows/Tokendial.Core/Install/InstallRecipe.cs
namespace Tokendial.Core.Install;

public enum InstallKind
{
    /// <summary>A command-line tool: the terminal chains its sign-in right after installing.</summary>
    Cli,
    /// <summary>A desktop app: signing in means opening it.</summary>
    App
}

/// <summary>Where a tool shows up once installed: executable names looked up on PATH, and absolute locations with ~, %VAR% and one * per segment.</summary>
public sealed record Detect(IReadOnlyList<string> Commands, IReadOnlyList<string> Paths)
{
    public static readonly Detect None = new([], []);
}

/// <summary>The tool's own sign-in command and the catalogue key of the sentence shown before it runs.</summary>
public sealed record SignInStep(string Command, string Hint);

/// <summary>One operating system's recipe: how to tell the tool is there, the vendor's install line, and how it signs in.</summary>
public sealed record PlatformRecipe(InstallKind Kind, Detect Detect, IReadOnlyList<string> Requires, string Install, SignInStep? SignIn, string? DownloadUrl)
{
    public bool NeedsWinget => Requires.Contains("winget");
    public bool NeedsBrew => Requires.Contains("brew");
}

/// <summary>
/// The install block of one provider spec in docs/providers, embedded in this assembly and never fetched
/// from the network.
/// </summary>
/// <remarks>
/// A map rather than one field per operating system, because the pair it replaced meant "everything that is
/// not Windows is macOS" - and so the assistant offered Homebrew commands to a Mint user. A platform that is
/// absent is absent: the assistant has nothing to offer rather than something wrong.
/// </remarks>
public sealed record InstallRecipe(string ProviderId, string Vendor, string DocsUrl, IReadOnlyDictionary<Desktop, PlatformRecipe> Platforms)
{
    public PlatformRecipe? On(Desktop desktop) => Platforms.GetValueOrDefault(desktop);

    public PlatformRecipe? Here => On(Roots.Current);

    public PlatformRecipe Windows => Platforms[Desktop.Windows];
    public PlatformRecipe MacOS => Platforms[Desktop.MacOS];
    public PlatformRecipe? Linux => On(Desktop.Linux);
}

File: macos/TokendialCore/Sources/TokendialCore/Install/InstallRecipe.swift
import Foundation

/// How signing in works once the tool is there: a cli takes its sign-in command in the same terminal, an app
/// is signed into in its own window.
public enum InstallKind: String, Sendable {
    case cli
    case app
}

/// Where a tool can be found once it is installed.
public struct Detect: Equatable, Sendable {
    public var commands: [String]
    public var paths: [String]

    public init(commands: [String] = [], paths: [String] = []) {
        self.commands = commands
        self.paths = paths
    }

    public static let none = Detect()
}

public struct SignInStep: Equatable, Sendable {
    public var command: String
    /// Catalogue key of the sentence shown before the command runs.
    public var hint: String

    public init(command: String, hint: String) {
        self.command = command
        self.hint = hint
    }
}

/// One provider's recipe. The shared spec carries a Windows half as well; this core reads only its own, so
/// there is nothing here to pick between at runtime.
public struct InstallRecipe: Equatable, Sendable {
    public var providerId: String
    public var vendor: String
    public var docsUrl: String
    public var kind: InstallKind
    public var detect: Detect
    public var requires: [String]
    public var install: String
    public var signIn: SignInStep?
    public var downloadUrl: String?

    public init(providerId: String, vendor: String, docsUrl: String, kind: InstallKind, detect: Detect,
                requires: [String], install: String, signIn: SignInStep?, downloadUrl: String?) {
        self.providerId = providerId
        self.vendor = vendor
        self.docsUrl = docsUrl
        self.kind = kind
        self.detect = detect
        self.requires = requires
        self.install = install
        self.signIn = signIn
        self.downloadUrl = downloadUrl
    }

    public var needsBrew: Bool { requires.contains("brew") }
    public var needsNpm: Bool { requires.contains("npm") }
}
